---
title: "mirdeep_analysis"
author: "Brecia Douglas"
date: "2024-02-02"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, error = FALSE, fig.path='Figures/', dev=c('png', 'pdf'))
options(knitr.table.format = 'markdown')
options(scipen=10000)
```

Loading in R packages
```{r}
library(tidyverse)
library(dplyr)
library(kableExtra)
library(seqinr)
library(stringdist)
library(purrr)
library(ape)
library(msa)
library(ggplot2)
```

Read in mirdeep files, bonafides and find shared miRNAs
```{r}
#read in bonafides file
bonafides_acerv <- read_csv("~/Desktop/mirna_1Nov/bonafides_master.csv")

all_mirdeep_files <- list.files(path = '~/Desktop/miRNA_Characterization/miRNA_Characterization/mirdeep_files/', full.names = TRUE) %>%
  tibble(file = .) %>%
  mutate(species = str_remove(file, dirname(file)) %>% str_remove('//') %>% str_remove('result_') %>% str_remove('.csv')) %>%
  rowwise(species) %>%
  summarise(read_csv(file), .groups = 'drop') %>%
  filter(provisional_id %in% bonafides_acerv$provisional_id)

#shared mirnas according to mirdeep seed
shared_mirnas <- all_mirdeep_files %>%
  select("species", "provisional_id", "example_miRBase_miRNA_with_the_same_seed") %>%
  filter(example_miRBase_miRNA_with_the_same_seed != "-") %>%
  pivot_wider(names_from = species, values_from = example_miRBase_miRNA_with_the_same_seed) %>%
  filter(provisional_id %in% bonafides_acerv$provisional_id)
kable(shared_mirnas)

common_mirnas <- shared_mirnas %>%
  mutate_all(funs(replace(., is.na(.), "-"))) %>%
  mutate(n_missing = apply(. == "-", 1, sum)) %>%
  filter(n_missing <= 8) 

kable(common_mirnas)
```

Getting conserved miRNAs from mature sequences using criteria (Wheeler et al., 2009)
1. nt 2-7 are exact match -> seed
2. length is +/- 2
3. remainder of alignment only can contain up to 3 mismatches
```{r}
fasta_to_table <- function(input_file) {
  input_file %>%
  as_tibble() %>%
  dplyr::rename(variable = V1) %>%
  group_by(grp = str_c('variable', rep(1:2, length.out = n()))) %>%
  mutate(rn = row_number()) %>%
  ungroup %>%
  pivot_wider(names_from = grp, values_from = variable) %>%
  select(-rn) %>%
  dplyr::rename(mirna_id = variable1, mature_sequence = variable2) %>%
  separate(col=mirna_id, into=c("carrot", "mirna_id"), sep=">") %>%
  select(-carrot)
}

mirna_seq_files <- list.files(path = '~/Desktop/miRNA_Characterization/miRNA_Characterization/raw_mirna_seq_cnidarians/', full.names = TRUE) %>%
  tibble(file = .) %>%
  mutate(species = str_remove(file, dirname(file)) %>% 
           str_remove('//') %>% str_remove('_mirnas_t.fa') %>% str_remove('_t.fa') %>% str_remove('praher_')) %>%
  rowwise(species) %>%
  summarise(fasta_to_table(read.table(file)), .groups = 'drop') %>%
  mutate(seed_sequence = substr(mature_sequence, 2, 7)) %>%
  mutate(leftover_seq = str_remove(mature_sequence, seed_sequence)) %>%
  mutate(mirna_len = str_length(mature_sequence))

get_conserved <- function(specie, file) {
  target_species <- file %>% filter(species == specie)

  other_mirnas <- file %>% filter(species != specie)
  
  seed_seq <- target_species$seed_sequence
  len_seq <- target_species$mirna_len
  leftover_seq <- target_species$leftover_seq
  
  #nt 2-7 of seed match
  find_seed <- function(seed){match_data <- filter(other_mirnas, seed_sequence == seed)}
  
  #length is within +/- 2
  find_len <- function(mir_length){match_data <- filter(other_mirnas, mirna_len <= mir_length + 2 & mirna_len >= mir_length - 2)}
  
  #no more than 3 mismatches in sequence without seed
  find_lev <- function(leftover){match_data <- filter(other_mirnas, stringdist(leftover, leftover_seq, method = "lv") < 3)}

  match_list_seed <- lapply(seed_seq, find_seed)
  match_list_len <- lapply(len_seq, find_len)
  match_list_left <- lapply(leftover_seq, find_lev)
  
  target_species$seed_matches <- match_list_seed
  target_species$len_matches <- match_list_len
  target_species$lev_good <- match_list_left

  targets_species_final <- target_species %>%
    mutate(temp = map2(seed_matches, len_matches, inner_join)) %>%
    mutate(conserved = map2(temp, lev_good, inner_join)) %>%
    select(-temp)
}

stylophora <- get_conserved("spi", mirna_seq_files)
cervicornis <- get_conserved("ace", mirna_seq_files)
millepora <- get_conserved("ami", mirna_seq_files)
digitifera <- get_conserved("adi", mirna_seq_files)
nematostella <- get_conserved("nve", mirna_seq_files)
exaiptsia <- get_conserved("epa", mirna_seq_files)
carnea <- get_conserved("eca", mirna_seq_files)
callimorphus <- get_conserved("sca", mirna_seq_files)
senile <- get_conserved("mse", mirna_seq_files)
viridis <- get_conserved("avi", mirna_seq_files)
jardinei <- get_conserved("cja", mirna_seq_files)
coerulea <- get_conserved("hco", mirna_seq_files)
aurita <- get_conserved("aau", mirna_seq_files)
malayensis <- get_conserved("sma", mirna_seq_files)
esculentum <- get_conserved("res", mirna_seq_files)
hydra <- get_conserved("hma", mirna_seq_files)

all_species <- rbind(stylophora, cervicornis, millepora ,digitifera, nematostella , exaiptsia, carnea, callimorphus, senile, viridis, jardinei, coerulea, aurita, malayensis, esculentum, hydra) %>%
  filter(map_lgl(conserved, ~ nrow(.) != 0))

all_species_shared <- all_species %>%
  select(-seed_matches, -len_matches, -lev_good) %>%
  unnest(cols = conserved, names_sep = "_") %>%
  mutate(base_call = paste(species, mirna_id, sep = "_")) %>%
  mutate(conserved_call = paste(conserved_species, conserved_mirna_id, sep = "_")) %>%
  select(base_call, conserved_call) %>%
  pivot_wider(names_from = base_call, values_from = conserved_call, values_fill = NA, values_fn = list) %>%
  pivot_longer(everything()) %>% 
  mutate(value = map(value, `length<-`, max(lengths(value)))) %>% 
  pivot_wider(names_from = name, values_from = value) %>% 
  unnest(everything())
```

Base tree that is then edited in FigTree
```{r}
myTree <- ape::read.tree(text = "((Bilateria), (((Hma),((Sma, Aau), Res)), ((Hco), ((((Cja), ((Spi), ((Ace), (Adi, Ami))))), (((Avi),(Epa, Mse)), (Eca, Sca, Nve))))));")
plot(myTree)

ape::write.nexus(myTree, file='~/Desktop/miRNA_Characterization/miRNA_Characterization/base_cnidarian_tree.nex')
```

Showing common miRNAs alignment
```{r results='asis'}
mirna_100 <- readDNAStringSet("~/Desktop/miRNA_Characterization/miRNA_Characterization/common_mirnas_fa/mir100.fa", format="fasta")
mirna_2030 <- readDNAStringSet("~/Desktop/miRNA_Characterization/miRNA_Characterization/common_mirnas_fa/mir2030.fa", format="fasta")
mirna_2022 <- readDNAStringSet("~/Desktop/miRNA_Characterization/miRNA_Characterization/common_mirnas_fa/mir2022.fa", format="fasta")
mirna_2023 <- readDNAStringSet("~/Desktop/miRNA_Characterization/miRNA_Characterization/common_mirnas_fa/mir2023.fa", format="fasta")
mirna_2037 <- readDNAStringSet("~/Desktop/miRNA_Characterization/miRNA_Characterization/common_mirnas_fa/mir2037.fa", format="fasta")
mirna_2036 <- readDNAStringSet("~/Desktop/miRNA_Characterization/miRNA_Characterization/common_mirnas_fa/mir2036.fa", format="fasta")
mirna_2025 <- readDNAStringSet("~/Desktop/miRNA_Characterization/miRNA_Characterization/common_mirnas_fa/mir2025.fa", format="fasta")

mir100_aln <- msa(mirna_100)
mir2030_aln <- msa(mirna_2030)
mir2022_aln <- msa(mirna_2022)
mir2023_aln <- msa(mirna_2023)
mir2037_aln <- msa(mirna_2037)
mir2036_aln <- msa(mirna_2036)
mir2025_aln <- msa(mirna_2025)

msaPrettyPrint(mir100_aln, output="asis", showNames="right", showNumbering = "none", showConsensus = "bottom", consensusColors = "RedBlue", logoColors = "standard area", furtherCode=c("\\defconsensus{.}{lower}{upper}"), showLegend = FALSE, askForOverwrite = FALSE)

msaPrettyPrint(mir2030_aln, output="asis", showNames="right", showNumbering = "none", showConsensus = "bottom", consensusColors = "RedBlue", logoColors = "standard area", furtherCode=c("\\defconsensus{.}{lower}{upper}"), showLegend = FALSE, askForOverwrite = FALSE)

msaPrettyPrint(mir2022_aln, output="asis", showNames="right", showNumbering = "none", showConsensus = "bottom", consensusColors = "RedBlue", logoColors = "standard area", furtherCode=c("\\defconsensus{.}{lower}{upper}"), showLegend = FALSE, askForOverwrite = FALSE)

msaPrettyPrint(mir2023_aln, output="asis", showNames="right", showNumbering = "none", showConsensus = "bottom", consensusColors = "RedBlue", logoColors = "standard area", furtherCode=c("\\defconsensus{.}{lower}{upper}"), showLegend = FALSE, askForOverwrite = FALSE)

msaPrettyPrint(mir2037_aln, output="asis", showNames="right", showNumbering = "none", showConsensus = "bottom", consensusColors = "RedBlue", logoColors = "standard area", furtherCode=c("\\defconsensus{.}{lower}{upper}"), showLegend = FALSE, askForOverwrite = FALSE)

msaPrettyPrint(mir2036_aln, output="asis", showNames="right", showNumbering = "none", showConsensus = "bottom", consensusColors = "RedBlue", logoColors = "standard area", furtherCode=c("\\defconsensus{.}{lower}{upper}"), showLegend = FALSE, askForOverwrite = FALSE)

msaPrettyPrint(mir2025_aln, output="asis", showNames="right", showNumbering = "none", showConsensus = "bottom", consensusColors = "RedBlue", logoColors = "standard area", furtherCode=c("\\defconsensus{.}{lower}{upper}"), showLegend = FALSE, askForOverwrite = FALSE)
```

Acropora cervicornis length distribution graph
```{r}
acerv_mirnas <- all_mirdeep_files %>%
  filter(species == "adi") %>%
  mutate(mirna_length = str_length(consensus_mature_sequence)) %>%
  rowwise() %>%
  mutate(count_u = str_count(consensus_mature_sequence,"u")) %>%
  mutate(count_g = str_count(consensus_mature_sequence,"g")) %>%
  mutate(count_c = str_count(consensus_mature_sequence,"c")) %>%
  mutate(count_a = str_count(consensus_mature_sequence,"a")) %>%
  mutate(start_nt = str_sub(consensus_mature_sequence, 1, 1)) %>%
  select(mirna_length, start_nt) %>%
  table() %>%
  as.data.frame()

ggplot(acerv_mirnas, aes(x=mirna_length, y=Freq, fill=start_nt)) +
          geom_bar(stat="identity") +
  xlab("Read Length") +
  ylab("Percent of Total Reads") +
  theme(legend.title= element_blank(),
        legend.position = c(.95, .95),
    legend.justification = c("right", "top"),
    legend.box.just = "right",
    legend.margin = margin(6, 6, 6, 6)) 
```


